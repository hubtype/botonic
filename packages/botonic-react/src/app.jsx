export class App {
  constructor() {
    throw 'This setup is deprecated in Botonic versions >=0.9.0. Please read the migration guide at: https://github.com/hubtype/botonic/tree/master/packages/botonic-react/MIGRATION_GUIDE.md'
  }
}
