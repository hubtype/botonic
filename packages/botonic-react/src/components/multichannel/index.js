import { Multichannel } from './multichannel'
import { MultichannelText } from './m-text'
import { MultichannelButton } from './m-button'
import { MultichannelReply } from './m-reply'
import { MultichannelCarousel } from './m-carousel'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
