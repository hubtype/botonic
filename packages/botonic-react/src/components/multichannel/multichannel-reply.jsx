import React, { useContext } from 'react'

import { RequestContext } from '../../contexts'
import { Reply } from '../reply'
import { isWhatsapp } from './multichannel-utils'

export const MultichannelReply = props => {
  const requestContext = useContext(RequestContext)
  const hasPath = () => Boolean(props.path)
  const hasPayload = () => Boolean(props.payload)
  const getText = () => `${props.children}`

  if (isWhatsapp(requestContext)) {
    if (hasPath() || hasPayload()) return `${getText()}`
    return null
  } else {
    return <Reply {...props}>{props.children}</Reply>
  }
}
