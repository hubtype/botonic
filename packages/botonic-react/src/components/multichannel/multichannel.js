import React from 'react'
import { RequestContext } from '../../contexts'

export class Multichannel extends React.Component {
  static contextType = RequestContext
  constructor(props) {
    super(props)
  }
  render() {
    if (this.context.session.user.provider == this.props.channel) {
      return this.props.children
    }
    return null
  }
}
