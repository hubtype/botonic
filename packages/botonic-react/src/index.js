export { BotonicInputTester, BotonicOutputTester } from './botonic-tester'
export * from './components'
export { RequestContext, WebchatContext } from './contexts'
export { DevApp } from './dev-app'
export { msgsToBotonic, msgToBotonic } from './msg-to-botonic'
export { NodeApp } from './node-app'
export { staticAsset } from './util/environment'
export { getBotonicApp, Webchat } from './webchat/index.js'
export { WebchatApp } from './webchat-app'
export { WebviewApp } from './webview'
