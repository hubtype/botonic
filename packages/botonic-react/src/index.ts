const React = require('react')
const ReactDOMServer = require('react-dom/server')

import { default as i18n } from '@botonic/core/lib/i18n'

declare global {
    namespace JSX {
        interface IntrinsicElements {
            [elemName: string]: any;
        }
    }
}

export async function renderReactAction(req, action) {
	let props = {}
	if(action.botonicInit) {
       props = await action.botonicInit({req})
    }

    let component = React.createElement(action, (props && typeof(props)==='object') ? Object.assign(props, req) : req)
    return ReactDOMServer.renderToStaticMarkup(component)
}

export function setLocale(req: any, locale: string) {
    req.context.__locale = locale
    i18n.setLocale(locale)
}
export { BotonicWebview } from './webview'
export { MessageTemplate } from './messageTemplate'
export { ShareButton } from './shareButton'
