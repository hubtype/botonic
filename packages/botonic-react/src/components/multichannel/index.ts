import { Multichannel } from './multichannel'
import { MultichannelButton } from './multichannel-button'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelReply } from './multichannel-reply'
import { MultichannelText } from './multichannel-text'

export { Multichannel }
export { MultichannelText }
export { MultichannelButton }
export { MultichannelReply }
export { MultichannelCarousel }
