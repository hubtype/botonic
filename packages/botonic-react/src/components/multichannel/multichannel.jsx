import React, { useContext } from 'react'

import { COMPONENT_TYPE } from '../../constants'
import { RequestContext } from '../../contexts'
import { deepMapWithIndex } from '../../util/react'
import { Text } from '../text'
import { MultichannelButton } from './multichannel-button'
import { MultichannelCarousel } from './multichannel-carousel'
import { MultichannelContext } from './multichannel-context'
import { MultichannelReply } from './multichannel-reply'
import { MultichannelText } from './multichannel-text'
import {
  isFacebook,
  isWhatsapp,
  MULTICHANNEL_WHATSAPP_PROPS,
} from './multichannel-utils'

export const Multichannel = props => {
  const requestContext = useContext(RequestContext)

  if (!isWhatsapp(requestContext) && !isFacebook(requestContext)) {
    return props.children
  }
  if (isFacebook(requestContext)) {
    const newChildren = deepMapWithIndex(props.children, child => {
      if (child && child.type && child.type.name === COMPONENT_TYPE.TEXT) {
        return (
          <MultichannelText {...child.props} key={child.key}>
            {child.props.children}
          </MultichannelText>
        )
      }
      return child
    })
    return newChildren
  }

  let newChildren = deepMapWithIndex(props.children, (child, index) => {
    if (child && child.type && child.type.name === COMPONENT_TYPE.BUTTON) {
      return (
        <MultichannelButton {...child.props} key={child.key}>
          {child.props.children}
        </MultichannelButton>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.REPLY) {
      return (
        <MultichannelReply {...child.props} key={child.key}>
          {child.props.children}
        </MultichannelReply>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.TEXT) {
      return (
        <MultichannelText
          {...child.props}
          {...props.text}
          key={child.key}
          {...(props.messageSeparator &&
            index > 0 && { newline: props.messageSeparator })}
        >
          {child.props.children}
        </MultichannelText>
      )
    }
    if (child && child.type && child.type.name === COMPONENT_TYPE.CAROUSEL) {
      return (
        <MultichannelCarousel
          {...child.props}
          {...props.carousel}
          key={child.key}
        >
          {child.props.children}
        </MultichannelCarousel>
      )
    }
    return child
  })
  if (props.messageSeparator != null) {
    newChildren = newChildren.map((c, index) =>
      index > 0 && typeof c === 'string' ? props.messageSeparator + c : c
    )
    newChildren = (
      <Text key={props.key} {...MULTICHANNEL_WHATSAPP_PROPS}>
        {newChildren}
      </Text>
    )
  }
  return (
    <MultichannelContext.Provider
      value={{
        currentIndex: props.firstIndex,
        boldIndex: props.boldIndex,
        indexSeparator: props.indexSeparator,
        messageSeparator: props.messageSeparator,
      }}
    >
      {newChildren}
    </MultichannelContext.Provider>
  )
}
