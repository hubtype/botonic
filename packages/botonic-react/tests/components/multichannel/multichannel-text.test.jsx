import { Text } from '../../../src'
import {
  MultichannelText,
  MultichannelButton,
} from '../../../src/components/multichannel'
import React from 'react'
import { whatsappRenderer } from '../../helpers/test-utils'
import { MultichannelReply } from '../../../src/components/multichannel/multichannel-reply'

describe('Multichannel text', () => {
  test('just text', () => {
    const text = <Text>Some text</Text>
    const sut = (
      <MultichannelText {...text.props}>{text.props.children}</MultichannelText>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with buttons plain', () => {
    const text = (
      <Text>
        The verbose text
        <MultichannelButton key='0' payload='payload1'>
          button text1
        </MultichannelButton>
        <MultichannelButton key='1' url='http://adrss'>
          button text2
        </MultichannelButton>
        <MultichannelButton key='2' path='path'>
          button text3
        </MultichannelButton>
      </Text>
    )
    const sut = (
      <MultichannelText {...text.props}>{text.props.children}</MultichannelText>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with buttons as array', () => {
    const text = (
      <Text>
        The verbose text
        {[
          <MultichannelButton key={'0'} payload={'payload1'}>
            button text1
          </MultichannelButton>,
          <MultichannelButton key={'1'} url='http://adrss'>
            button text2
          </MultichannelButton>,
          <MultichannelButton key={'2'} path='path'>
            button text3
          </MultichannelButton>,
        ]}
      </Text>
    )

    const sut = (
      <MultichannelText {...text.props}>{text.props.children}</MultichannelText>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with replies plain', () => {
    const text = (
      <Text>
        Quick replies
        <MultichannelReply key='0' payload='payload1'>
          reply text1
        </MultichannelReply>
        <MultichannelReply key='2' path='path'>
          reply text2
        </MultichannelReply>
      </Text>
    )
    const sut = (
      <MultichannelText {...text.props}>{text.props.children}</MultichannelText>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with replies as array', () => {
    const text = (
      <Text>
        The verbose text
        {[
          <MultichannelReply key='0' payload='payload1'>
            reply text1
          </MultichannelReply>,
          <MultichannelReply key='2' path='path'>
            reply text2
          </MultichannelReply>,
        ]}
      </Text>
    )

    const sut = (
      <MultichannelText {...text.props}>{text.props.children}</MultichannelText>
    )
    const renderer = whatsappRenderer(sut)
    const tree = renderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
