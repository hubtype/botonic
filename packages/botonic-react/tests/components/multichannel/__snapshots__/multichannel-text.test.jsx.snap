// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multichannel text just text 1`] = `
<message
  delay={0}
  indexMode="number"
  markdown={false}
  type="text"
  typing={0}
>
  Some text
</message>
`;

exports[`Multichannel text with buttons as array 1`] = `
<message
  delay={0}
  indexMode="number"
  markdown={false}
  type="text"
  typing={0}
>
  The verbose text
  
1. button text1
  
2. button text3
  
- button text2: http://adrss
</message>
`;

exports[`Multichannel text with buttons plain 1`] = `
<message
  delay={0}
  indexMode="number"
  markdown={false}
  type="text"
  typing={0}
>
  The verbose text
  
1. button text1
  
2. button text3
  
- button text2: http://adrss
</message>
`;

exports[`Multichannel text with replies as array 1`] = `
<message
  delay={0}
  indexMode="number"
  markdown={false}
  type="text"
  typing={0}
>
  The verbose text
  
1. reply text1
  
2. reply text2
</message>
`;

exports[`Multichannel text with replies plain 1`] = `
<message
  delay={0}
  indexMode="number"
  markdown={false}
  type="text"
  typing={0}
>
  Quick replies
  
1. reply text1
  
2. reply text2
</message>
`;
